core/base/test/TSystemCompileMacro.cxx

```diff
@@ -0,0 +1,116 @@
+#include <cstdio>
+#include <fstream>
+
+#include <string>
+#include <thread>
+#include <vector>
+
+#include <gtest/gtest.h>
+
+#include <ROOT/TestSupport.hxx>
+
+#include "TROOT.h"
+#include "TSystem.h"
+
+#if defined(_WIN32)
+const char *suffix = ".dll";
+#else
+const char *suffix = ".so";
+#endif
+
+const static std::string code_template{R"(
+#ifndef TSYSTEMCOMPILEMACROTHREADSAFETY_HELPER_%s
+#define TSYSTEMCOMPILEMACROTHREADSAFETY_HELPER_%s
+bool %s_%s(double x, double val)
+{
+    return x > val;
+}
+#endif
+)"};
+
+void write_n_helpers(const std::string &basename, unsigned int n)
+{
+   for (unsigned int i = 0; i < n; i++) {
+      char buffer[300];
+      const auto itos = std::to_string(i);
+      std::snprintf(buffer, 300, code_template.c_str(), itos.c_str(), itos.c_str(), basename.c_str(), itos.c_str());
+      std::string library_file{basename + "_" + itos + ".cpp"};
+      std::ofstream myfile{library_file};
+      myfile << buffer;
+   }
+}
+
+void compile_macro_n_times(const std::string &filename, unsigned int n)
+{
+   for (unsigned int i = 0; i < n; i++) {
+      gSystem->CompileMacro(filename.c_str(), "O");
+   }
+}
+
+void remove_n_library_artifacts(const std::string &basename, unsigned int n)
```

Comment:
Suppose a crash happens before the function is invoked and the artifacts are not removed. Wouldn't perhaps RAII help?